README.md
SlackMessage: a Friendly DSL for Slack
=============

SlackMessage is a wrapper over the [Block Kit
API](https://app.slack.com/block-kit-builder/) to make it easy to read and
write messages to slack in your ruby application.

Posting a message to Slack should be this easy:

```ruby
SlackMessage.post_to('#general') do
  text "We did it @here! :thumbsup:"
end
```
 
### The Docs

You'll find much more information about how to use SlackMessage by visiting
[the docs](https://jmmastey.github.io/slack_message).
 

### A Rich DSL Focused on Maintainability

SlackMessage is able to build all kinds of rich messages for you. It focuses on
writing code that looks similar to the output messages themselves, with as
little repetition and cruft as possible.

```ruby
SlackMessage.post_to('#general') do
  section do
    text "A job has generated some output for you to review."
    text 'And More' * 10
    link_button "See Results", "https://google.com"
  end

  section do
    text ":unlock-new: New Data Summary"

    list_item "Date", "09/05/2021"
    list_item "Total Imported", 45_004
    list_item "Total Errors", 5
  end

  divider

  section do
    text "See more here: #{link('result', 'https://google.com')}"
  end

  text ":rocketship: <hello@joemastey.com>"

  context ":custom_slack_emoji: An example footer *with some markdown*."
end
```

### Opinionated Stances

This gem is intended to stay simple. Other Slack gems have lots of config
options and abilities, which makes them powerful, but makes them a pain to use.

Accordingly, SlackMessage is developed with some strong opinions in mind:

* SlackMessage has no dependencies. Your lockfile is enough of a mess already.
* A DSL that focuses on code that is easy to write, read, and maintain. Only
  features that can be implemented with that in mind are included.
* The code to build a message should look a lot like the message itself. Code
  that is simple to read and understand is a priority.
* Webhooks are passé. Only Slack Apps are supported now.
* Unless you request otherwise, text is always rendered using `mrkdwn`. If you
  want plaintext, you'll need to ask for it. Same for the `emoji` flag.
* As many API semantics as possible are hidden. For instance, if you post to
  something that looks like an email address, `slack_message` is going to try
  to look it up as an email address.
* A few little hacks on the block syntax, such as adding a `blank_line` (which
  doesn't exist in the API), or leading spaces.
* Configuration is kept simple, with helpers for frequently reused bots.

Some changes that are still planned or desired, but not yet added:

* any interactive elements at all: https://api.slack.com/interactivity/handling
* multiple recipients: https://api.slack.com/methods/conversations.open
* more mrkdwn syntax, like quotes or code blocks https://api.slack.com/reference/surfaces/formatting#line-breaks
* more and better organized testing capability (for scheduled messages, editing, deleting)
* posting ephemeral messages: https://api.slack.com/methods/chat.postEphemeral
* easier way to dump / load message responses
* updated docs w/ links to BlockBuilder

### Contributing

Contributions are very welcome. Fork, fix, submit pull. Since simplicity of API
is a strong priority, so opening an issue to discuss possible interface changes
would be wise.

Contribution is expected to conform to the [Contributor
Covenant](https://github.com/jmmastey/slack_message/blob/master/CODE_OF_CONDUCT.md).

### License

This software is released under the [MIT
License](https://github.com/jmmastey/slack_message/blob/master/MIT-LICENSE).


CHANGELOG.md
# Changelog

## [3.4.0] - 2023-02-21
- Add full support for ruby 3.0.x kwargs

## [3.3.0] - 2022-07-21
- Differentiate errors for bad JSON versus invalid block data.
- Expand tests to cover many more cases.
- Add validation for block size limit in Slack API.
- Add CI check to prevent checkin of `fit`, `fcontext` etc.

## [3.2.0] - 2022-06-23
- Fix bugs introduced by accidental checkin of incomplete refactor.

## [3.1.0] - 2022-04-18
- Methods from the calling context can now be called within a section block.

## [3.0.2] - 2022-04-16
- Fix tests on ruby 3.0.
- More adjustments and additions to docs.
- Add warnings when overriding notification text and context block.

## [3.0.1] - 2021-12-22
- Major overhaul of error handling and expansion on which errors trigger
  friendly messages for users.
- More additions to the docs, and working github pages integration.
- It's my birthday!

## [3.0.0] - 2021-12-19
- Return a more structured object from successful message sends.
- Add the ability to edit or delete a message.
- Complete overhaul of docs because they were too large.

## [2.4.0] - 2021-12-13
- Add ability to schedule messages, plus some guard rails around that.
- Add ability to debug by logging out the total set of params sent to the API.

## [2.3.1] - 2021-11-30
- Adjust that minimum version by changing some syntax to older styles. Given
  support for ruby 2.4 ended almost 2 years ago, going to go ahead and leave
  it behind.
- Remove lockfile from repo

## [2.3.0] - 2021-11-30
- Formally require minimum version of ruby. It wouldn't have worked anyway,
  but worth actually specifying.

## [2.2.2] - 2021-11-30
- Add github workflow for automatic CI runs. Stolen from another project.

## [2.2.1] - 2021-11-20
- Trying to fetch user ID for a string that isn't email-like raises an error.
- In tests, fetching user IDs is mocked out to prevent network requests.
- Tightened up and clarified README.
- Some internal cleanup and restructuring of modules.

## [2.2.0] - 2021-11-20
- When sending text, it is now possible to mention users and have their user
  IDs automatically converted using `<email@email.com>` within text nodes.
- It's now possible to override notification text.
- Errors received from the Slack API now raise `SlackMessage::ApiError`.
- Re-exposed a top-level method for getting user IDs, `SlackMessage.user_id`.
- Raising some better errors when no message payload is present.
- Using `build` now requires a profile, so configuration must exist.

## [2.1.0] - 2021-11-01
- Change to use Slack Apps for all profiles. This should allow growth toward
  updating messages, working with interactive messages etc.
- As a result, allow custom icons per profile / message.
- When sending a message, the first `text` block is used for the notification
  content. Should resolve "this content cannot be displayed".
- Significant restructuring of README.

## [2.0.0] - 2021-11-01
- Yeah that was all broken.

## [1.9.0] - 2021-10-27
- Add many validations so that trying to add e.g. empty text won't succeed.
  Previously that would be accepted but return `invalid_blocks` from the API.

## [1.8.1] - 2021-10-08
- Cleaned that rspec code a bit, added more matchers for real world use.

## [1.8.0] - 2021-10-07
- Added the ability to test in RSpec.

## [1.7.1] - 2021-10-06
- Fixed literally a syntax issue.
- Fixed specs.
- Fixed API to include JSON since consumers may not have loaded it.

## [1.7.0] - 2021-10-06
- THIS RELEASE IS BADLY BROKEN.
- Added new error messages when API configuration is wrong / missing.
- Fixed issue with `instance_eval` and using methods within block.
- Fixed issue with sectionless `list_item`.

## [1.6.0] - 2021-10-04
- Added `:default_channel` and `post_as` to deal with repetitive channel usage.

## [1.5.0] - 2021-10-01
- Added `ol` and `ul` to sections w/ some formatting.

## [1.4.0] - 2021-09-27
- Changed `image` to `accessory_image` to differentiate between the image block
  and the accessory image within a block.

## [1.3.0] - 2021-09-27
- Added ability to use custom names when posting.
- Added ability to post images within sections.
- Added warnings for potentially invalid URLs.

## [1.2.0] - 2021-09-26
- Fixed gemspec, which was entirely broken.

## [1.1.0] - 2021-09-26
- Expanded the README significantly w/ usage instructions.
- Added lots of error handling to requests.

## [1.0.0] - 2021-09-25
- Added the base gem w/ a DSL for constructing blocks using sections.
- Added a changelog, apparently.


spec/spec_helper.rb
require_relative '../lib/slack_message'
require_relative '../lib/slack_message/rspec'

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration

RSpec.configure do |config|
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  if ENV['CI'] == "true"
    config.filter_run_when_matching :focus => lambda { raise ArgumentError, "Don't filter tests on checked in code." }
  else
    config.filter_run_when_matching :focus
  end

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed

  include SlackMessage::RSpec
end


spec/slack_message_spec.rb
require 'spec_helper'

RSpec.describe SlackMessage do

  describe "DSL" do
    describe "#build" do
      def outer_method
       "foo"
      end

      it "renders some JSON" do
        SlackMessage.configure do |config|
          config.clear_profiles!
          config.add_profile(name: 'default profile', api_token: 'abc123')
        end

        expected_output = [
          { type: "section",
            text: { text: "foo", type: "mrkdwn" }
          },
          { type: "section",
            text: { text: "foo", type: "mrkdwn" }
          },
        ]

        output = SlackMessage.build do
          text outer_method()
          section do
            text outer_method()
          end
        end

        expect(output).to eq(expected_output)
      end
    end
  end

  describe "configuration" do
    it "lets you add and fetch profiles" do
      SlackMessage.configure do |config|
        config.clear_profiles!
        config.add_profile(name: 'default profile', api_token: 'abc123')
        config.add_profile(:nonstandard, name: 'another profile', api_token: 'abc123')
      end

      expect(SlackMessage.configuration.profile(:default)[:name]).to eq('default profile')
      expect(SlackMessage.configuration.profile(:nonstandard)[:name]).to eq('another profile')

      expect {
        SlackMessage.configuration.profile(:missing)
      }.to raise_error(ArgumentError)
    end
  end

  describe "custom expectations" do
    before do
      SlackMessage.configure do |config|
        config.clear_profiles!
        config.add_profile(name: 'default profile', api_token: 'abc123')
      end
    end

    it "can assert expectations against posts" do
      expect {
        SlackMessage.post_to('#lieutenant') { text "foo" }
      }.not_to post_slack_message_to('#general')

      expect {
        SlackMessage.post_to('#general') { text "foo" }
      }.to post_slack_message_to('#general').with_content_matching(/foo/)
    end

    it "resets state properly" do
      expect {
        SlackMessage.post_to('#general') { text "foo" }
      }.to post_slack_message_to('#general')

      expect { }.not_to post_slack_message_to('#general')
    end

    it "lets you assert by profile name" do
      SlackMessage.configure do |config|
        config.clear_profiles!
        config.add_profile(:schmoebot, name: 'Schmoe', api_token: 'abc123', icon: ':schmoebot:', default_channel: '#schmoes')
      end

      expect {
        SlackMessage.post_as(:schmoebot) { text "foo" }
      }.to post_slack_message_to('#schmoes')

      expect {
        SlackMessage.post_as(:schmoebot) { text "foo" }
      }.to post_slack_message_as(:schmoebot)

      expect {
        SlackMessage.post_as(:schmoebot) { text "foo" }
      }.to post_slack_message_as('Schmoe').with_content_matching(/foo/)
    end

    it "lets you assert by profile image" do
      SlackMessage.configure do |config|
        config.clear_profiles!
        config.add_profile(:schmoebot, name: 'Schmoe', api_token: 'abc123', icon: ':schmoebot:', default_channel: '#schmoes')
      end

      expect {
        SlackMessage.post_as(:schmoebot) { text "foo" }
      }.to post_slack_message_with_icon(':schmoebot:')

      expect {
        SlackMessage.post_as(:schmoebot) do
          bot_icon ':schmalternate:'
          text "foo"
        end
      }.to post_slack_message_with_icon(':schmalternate:')

      expect {
        SlackMessage.post_as(:schmoebot) do
          bot_icon 'https://thispersondoesnotexist.com/image'
          text "foo"
        end
      }.to post_slack_message_with_icon_matching(/thisperson/)
    end

    it "can assert more generally too tbh" do
      expect {
        SlackMessage.post_to('#general') { text "foo" }
      }.to post_to_slack.with_content_matching(/foo/)
    end

    it "lets you send links with custom styles" do
      expect {
        SlackMessage.post_to('#general') do
          link_button 'Does this person exist?', 'https://thispersondoesnotexist.com/image', style: :danger
        end
      }.to post_to_slack.with_content_matching(/danger/)
    end
  end

  describe "API convenience" do
    before do
      SlackMessage.configure do |config|
        config.clear_profiles!
        config.add_profile(name: 'default profile', api_token: 'abc123')
      end
    end

    it "converts user IDs within text when tagged properly" do
      allow(SlackMessage::Api).to receive(:user_id_for).and_return('ABC123')

      expect {
        SlackMessage.post_to('#general') { text("Working: <hello@joemastey.com> ") }
      }.to post_to_slack.with_content_matching(/ABC123/)

      expect {
        SlackMessage.post_to('#general') { text("Not Tagged: hello@joemastey.com ") }
      }.to post_to_slack.with_content_matching(/hello@joemastey.com/)
    end

    it "is graceful about those failures" do
      allow(SlackMessage::Api).to receive(:user_id_for).with('nuffin@nuffin.nuffin', any_args).and_raise(SlackMessage::ApiError)
      allow(SlackMessage::Api).to receive(:user_id_for).with('hello@joemastey.com', any_args).and_return('ABC123')

      expect {
        SlackMessage.post_to('#general') { text("Not User: <nuffin@nuffin.nuffin>") }
      }.to post_to_slack.with_content_matching(/\<nuffin@nuffin.nuffin\>/)

      expect {
        SlackMessage.post_to('#general') { text("Not User: <nuffin@nuffin.nuffin>, User: <hello@joemastey.com>") }
      }.to post_to_slack.with_content_matching(/ABC123/)
    end
  end

  describe "error handling" do
    before do
      SlackMessage.configure do |config|
        config.clear_profiles!
        config.add_profile(name: 'default profile', api_token: 'abc123')
        config.add_profile(:schmoebot, name: 'Schmoe', api_token: 'abc123', icon: ':schmoebot:', default_channel: '#schmoes')
      end
    end

    after do
      SlackMessage::RSpec.reset_mock_response
    end

    it "raises for redirects" do
      SlackMessage::RSpec.respond_with(code: '302')

      expect {
        SlackMessage.post_to('#general') { text 'nuh uh' }
      }.to raise_error(SlackMessage::ApiError)
    end

    it "even raises errors during deletes" do
      message = SlackMessage.post_to('#general') { text 'nuh uh' }

      SlackMessage::RSpec.respond_with({'error' => 'bad choice'})

      expect {
        SlackMessage.delete(message)
      }.to raise_error(SlackMessage::ApiError)
    end

    shared_examples 'post api error message' do |error, error_message|
      it "responds to posts with error code '#{error}' with the expected message" do
        SlackMessage::RSpec.respond_with({'error' => error})

        expect {
          SlackMessage.post_to('#general') { text 'nuh uh' }
        }.to raise_error(SlackMessage::ApiError).with_message(error_message)

        expect {
          SlackMessage.post_as(:schmoebot) { text 'nuh uh' }
        }.to raise_error(SlackMessage::ApiError).with_message(error_message)
      end
    end

    shared_examples 'update api error message' do |error, error_message|
      it "responds to updates with error code '#{error}' with the expected message" do
        message = SlackMessage.post_to('#general') { text 'nuh uh' }

        SlackMessage::RSpec.respond_with({'error' => error})

        expect {
          SlackMessage.update(message) { text 'nuh uh' }
        }.to raise_error(SlackMessage::ApiError).with_message(error_message)
      end
    end

    shared_examples 'delete api error message' do |error, error_message|
      it "responds to updates with error code '#{error}' with the expected message" do
        message = SlackMessage.post_to('#general') { text 'nuh uh' }

        SlackMessage::RSpec.respond_with({'error' => error})

        expect {
          SlackMessage.delete(message)
        }.to raise_error(SlackMessage::ApiError).with_message(error_message)
      end
    end

    shared_examples 'block api error message' do |error, error_message|
      include_examples 'post api error message', error, error_message
      include_examples 'update api error message', error, error_message
    end

    include_examples 'block api error message', 'nuffin', /Received error response 'nuffin' from Slack/
    include_examples 'block api error message', 'invalid_blocks', /because the request contained invalid blocks:\n\[Enable debugging in configuration to view block data\.\]/
    include_examples 'block api error message', 'invalid_blocks_format', /because blocks is not a valid JSON object or doesn't match the Block Kit syntax:\n\[Enable debugging in configuration to view block data\.\]/
    include_examples 'block api error message', 'channel_not_found', /Slack message to non-existent channel or user/
    include_examples 'block api error message', 'invalid_auth', /because the API key for profile '.*' is wrong, or the app has insufficient permissions \(invalid_auth\)/
    include_examples 'block api error message', 'message_too_long', /but the message was too long/
    include_examples 'block api error message', 'invalid_arguments', /with invalid payload/
    include_examples 'block api error message', 'rate_limited', /because you've reached your rate limit/

    # Scheduling messages
    include_examples 'post api error message', 'invalid_time', /because you requested an invalid time/
    include_examples 'post api error message', 'time_in_past', /because you requested a time in the past \(or too close to now\)/
    include_examples 'post api error message', 'time_too_far', /because you requested a time more than 120 days in the future/

    # Updating messages
    include_examples 'update api error message', 'message_not_found', /but the message wasn't found/

    # Deleting messages
    include_examples 'delete api error message', 'nuffin', /Received error response 'nuffin' from Slack/
    include_examples 'delete api error message', 'invalid_scheduled_message_id', /Can't delete message because the ID was invalid, or the message has already posted/
    include_examples 'delete api error message', 'message_not_found', /but the message wasn't found/
    include_examples 'delete api error message', 'cant_delete_message', /Can't delete message because 'default' doesn't have permission to/
    include_examples 'delete api error message', 'compliance_exports_prevent_deletion', /Can't delete message because team compliance settings prevent it/
    include_examples 'delete api error message', 'invalid_auth', /because the API key for profile '.*' is wrong, or the app has insufficient permissions \(invalid_auth\)/
    include_examples 'delete api error message', 'rate_limited', /because you've reached your rate limit/
  end
end


lib/slack_message.rb
module SlackMessage
  require 'slack_message/response'
  require 'slack_message/dsl'
  require 'slack_message/error_handling'
  require 'slack_message/api'
  require 'slack_message/configuration'

  EMAIL_TAG_PATTERN = /<[^@ \t\r\n\<]+@[^@ \t\r\n]+\.[^@ \t\r\n]+>/
  EMAIL_PATTERN = /^\S{1,}@\S{2,}\.\S{2,}$/

  class ApiError < RuntimeError; end

  def self.configuration
    Configuration
  end

  def self.configure(&block)
    configuration.configure(&block)
  end

  def self.user_id(email, profile_name = :default)
    profile = Configuration.profile(profile_name)
    Api.user_id_for(email, profile)
  end

  def self.post_to(target, as: :default, at: nil, &block)
    profile = Configuration.profile(as)

    payload = Dsl.new(block, profile).tap do |instance|
      instance.instance_eval(&block)
    end

    target  = Api::user_id_for(target, profile) if target =~ EMAIL_PATTERN

    Api.post(payload, target, profile, at)
  end

  def self.post_as(profile_name, at: nil, &block)
    profile = Configuration.profile(profile_name)
    if profile[:default_channel].nil?
      raise ArgumentError, "Sorry, you need to specify a default_channel for profile #{profile_name} to use post_as"
    end

    target  = profile[:default_channel]
    payload = Dsl.new(block, profile).tap do |instance|
      instance.instance_eval(&block)
    end

    target  = Api::user_id_for(target, profile) if target =~ EMAIL_PATTERN

    Api.post(payload, target, profile, at)
  end

  def self.update(message, &block)
    unless message.is_a?(SlackMessage::Response)
      raise ArgumentError, "You must pass in a SlackMessage::Response to update a message"
    end

    if message.scheduled?
      raise ArgumentError, "Sorry, scheduled messages cannot be updated. You will need to delete the message and schedule a new one."
    end

    profile = Configuration.profile(message.profile_handle)
    payload = Dsl.new(block, profile).tap do |instance|
      instance.instance_eval(&block)
    end

    Api.update(payload, message, profile)
  end

  def self.delete(message)
    unless message.is_a?(SlackMessage::Response)
      raise ArgumentError, "You must pass in a SlackMessage::Response to delete a message"
    end

    if message.sent_to_user?
      raise ArgumentError, "It's not possible to delete messages sent directly to users."
    end

    profile = Configuration.profile(message.profile_handle)
    Api.delete(message, profile)
  end

  def self.build(profile_name = :default, &block)
    profile = Configuration.profile(profile_name)

    Dsl.new(block, profile).tap do |instance|
      instance.instance_eval(&block)
    end.send(:render)
  end
end


lib/slack_message/response.rb
class SlackMessage::Response
  attr_reader :channel, :timestamp, :profile_handle, :scheduled_message_id, :original_response

  def initialize(api_response, profile_handle)
    @original_response = JSON.parse(api_response.body)
    @ok = @original_response["ok"]
    @channel = @original_response["channel"]

    @timestamp = @original_response["ts"]
    @scheduled_message_id = @original_response["scheduled_message_id"]

    @profile_handle = profile_handle
  end

  def marshal_dump
    [ @profile_handle, @channel, @timestamp, @original_response, @ok, @original_response ]
  end

  def marshal_load(data)
    @profile_handle, @channel, @timestamp, @original_response, @ok, @original_response = data
  end

  def sent_to_user?
    channel =~ /^D.*/ # users are D for DM, channels start w/ C
  end

  def scheduled?
    !!scheduled_message_id
  end

  def inspect
    identifier = if scheduled?
      "scheduled_message_id=#{scheduled_message_id}"
    else
      "timestamp=#{timestamp}"
    end

    ok_msg = @ok ? "ok" : "error"

    "<SlackMessage::Response #{ok_msg} profile_handle=:#{profile_handle} channel=#{channel} #{identifier}>"
  end
end


lib/slack_message/configuration.rb
module SlackMessage::Configuration
  @@profiles = {}
  @@debug = false

  def self.reset
    @@profiles = {}
    @@debug = false
  end

  def self.configure
    yield self
  end

  ###

  def self.clear_profiles! # test harness, mainly
    @@profiles = {}
  end

  def self.add_profile(handle = :default, api_token:, name: nil, icon: nil, default_channel: nil)
    if @@profiles.include?(handle)
      warn "WARNING: Overriding profile '#{handle}' in SlackMessage config"
    end

    @@profiles[handle] = {
      handle: handle,
      api_token: api_token,
      name: name,
      icon: icon,
      default_channel: default_channel
    }
  end

  def self.profile(handle)
    unless @@profiles.include?(handle)
      raise ArgumentError, "Unknown SlackMessage profile '#{handle}'."
    end

    @@profiles[handle]
  end

  def self.debug
    @@debug = true
  end

  def self.debugging?
    @@debug
  end
end


lib/slack_message/error_handling.rb
class SlackMessage::ErrorHandling
  PERMISSIONS_ERRORS = ["token_revoked", "token_expired", "invalid_auth", "not_authed",
                        "team_access_not_granted", "no_permission", "missing_scope",
                        "not_allowed_token_type", "ekm_access_denied"]

  def self.raise_post_response_errors(response, params, profile)
    body  = JSON.parse(response.body)
    error = body.fetch("error", "")

    if error == "invalid_blocks"
      raise SlackMessage::ApiError, "Couldn't send Slack message because the request contained invalid blocks:\n#{blocks_message(params[:blocks])}"
    elsif error == "invalid_blocks_format"
      raise SlackMessage::ApiError, "Couldn't send Slack message because blocks is not a valid JSON object or doesn't match the Block Kit syntax:\n#{blocks_message(params[:blocks])}"
    elsif error == "channel_not_found"
      raise SlackMessage::ApiError, "Tried to send Slack message to non-existent channel or user '#{params[:channel]}'"

    # scheduling messages
    elsif error == "invalid_time"
      raise SlackMessage::ApiError, "Couldn't schedule Slack message because you requested an invalid time '#{params[:post_at]}'"
    elsif error == "time_in_past"
      raise SlackMessage::ApiError, "Couldn't schedule Slack message because you requested a time in the past (or too close to now) '#{params[:post_at]}'"
    elsif error == "time_too_far"
      raise SlackMessage::ApiError, "Couldn't schedule Slack message because you requested a time more than 120 days in the future '#{params[:post_at]}'"


    elsif PERMISSIONS_ERRORS.include?(error)
      raise SlackMessage::ApiError, "Couldn't send Slack message because the API key for profile '#{profile[:handle]}' is wrong, or the app has insufficient permissions (#{error})"
    elsif error == "message_too_long"
      raise SlackMessage::ApiError, "Tried to send Slack message, but the message was too long"
    elsif error == "invalid_arguments"
      raise SlackMessage::ApiError, "Tried to send Slack message with invalid payload"
    elsif ["rate_limited", "ratelimited"].include?(error)
      raise SlackMessage::ApiError, "Couldn't send Slack message because you've reached your rate limit"
    elsif response.code == "302"
      raise SlackMessage::ApiError, "Got 302 response while posting to Slack. Check your API key for profile '#{profile[:handle]}'"
    elsif response.code != "200"
      raise SlackMessage::ApiError, "Got an error back from the Slack API (HTTP #{response.code}):\n#{response.body}"
    elsif !(error.nil? || error == "")
      raise SlackMessage::ApiError, "Received error response '#{error}' from Slack during message posting:\n#{response.body}"
    end
  end

  def self.raise_update_response_errors(response, params, profile)
    body  = JSON.parse(response.body)
    error = body.fetch("error", "")

    if error == "invalid_blocks"
      raise SlackMessage::ApiError, "Couldn't update Slack message because the request contained invalid blocks:\n#{blocks_message(params[:blocks])}"
    elsif error == "invalid_blocks_format"
      raise SlackMessage::ApiError, "Couldn't update Slack message because blocks is not a valid JSON object or doesn't match the Block Kit syntax:\n#{blocks_message(params[:blocks])}"
    elsif error == "channel_not_found"
      raise SlackMessage::ApiError, "Tried to update Slack message to non-existent channel or user '#{params[:channel]}'"

    elsif error == "message_not_found"
      raise SlackMessage::ApiError, "Tried to update Slack message, but the message wasn't found (timestamp '#{params[:ts]}' for channel '#{params[:channel]}'"
    elsif error == "cant_update_message"
      raise SlackMessage::ApiError, "Couldn't update message because the message type isn't able to be updated, or #{profile[:handle]} isn't allowed to update it"
    elsif error == "edit_window_closed"
      raise SlackMessage::ApiError, "Couldn't update message because it's too old"


    elsif PERMISSIONS_ERRORS.include?(error)
      raise SlackMessage::ApiError, "Couldn't update Slack message because the API key for profile '#{profile[:handle]}' is wrong, or the app has insufficient permissions (#{error})"
    elsif error == "message_too_long"
      raise SlackMessage::ApiError, "Tried to update Slack message, but the message was too long"
    elsif error == "invalid_arguments"
      raise SlackMessage::ApiError, "Tried to update Slack message with invalid payload"
    elsif ["rate_limited", "ratelimited"].include?(error)
      raise SlackMessage::ApiError, "Couldn't update Slack message because you've reached your rate limit"
    elsif response.code == "302"
      raise SlackMessage::ApiError, "Got 302 response while updating a message. Check your API key for profile '#{profile[:handle]}'"
    elsif response.code != "200"
      raise SlackMessage::ApiError, "Got an error back from the Slack API (HTTP #{response.code}):\n#{response.body}"
    elsif !(error.nil? || error == "")
      raise SlackMessage::ApiError, "Received error response '#{error}' from Slack during message update:\n#{response.body}"
    end
  end

  def self.raise_delete_response_errors(response, message, profile)
    body  = JSON.parse(response.body)
    error = body.fetch("error", "")

    if error == "channel_not_found"
      raise SlackMessage::ApiError, "Tried to delete Slack message in non-existent channel '#{message.channel}'"

    elsif error == "invalid_scheduled_message_id"
      raise SlackMessage::ApiError, "Can't delete message because the ID was invalid, or the message has already posted (#{message.scheduled_message_id})"
    elsif error == "message_not_found"
      raise SlackMessage::ApiError, "Tried to delete Slack message, but the message wasn't found (timestamp '#{message.timestamp}' for channel '#{message.channel}')"
    elsif error == "cant_delete_message"
      raise SlackMessage::ApiError, "Can't delete message because '#{profile[:handle]}' doesn't have permission to"
    elsif error == "compliance_exports_prevent_deletion"
      raise SlackMessage::ApiError, "Can't delete message because team compliance settings prevent it"


    elsif PERMISSIONS_ERRORS.include?(error)
      raise SlackMessage::ApiError, "Couldn't delete Slack message because the API key for profile '#{profile[:handle]}' is wrong, or the app has insufficient permissions (#{error})"
    elsif ["rate_limited", "ratelimited"].include?(error)
      raise SlackMessage::ApiError, "Couldn't delete Slack message because you've reached your rate limit"
    elsif response.code == "302"
      raise SlackMessage::ApiError, "Got 302 response while deleting a message. Check your API key for profile '#{profile[:handle]}'"
    elsif response.code != "200"
      raise SlackMessage::ApiError, "Got an error back from the Slack API (HTTP #{response.code}):\n#{response.body}"
    elsif !(error.nil? || error == "")
      raise SlackMessage::ApiError, "Received error response '#{error}' from Slack during message delete:\n#{response.body}"
    end
  end

  def self.raise_user_lookup_response_errors(response, email, profile)
    begin
      payload = JSON.parse(response.body)
    rescue
      raise SlackMessage::ApiError, "Unable to parse JSON response from Slack API\n#{response.body}"
    end

    error = payload["error"]

    if error == "users_not_found"
      raise SlackMessage::ApiError, "Couldn't find a user with the email '#{email}'"


    elsif PERMISSIONS_ERRORS.include?(error)
      raise SlackMessage::ApiError, "Couldn't look up users because the API key for profile '#{profile[:handle]}' is wrong, or the app has insufficient permissions (#{error})"
    elsif error
      raise SlackMessage::ApiError, "Received error response from Slack during user lookup:\n#{response.body}"
    elsif response.code == "302"
      raise SlackMessage::ApiError, "Got 302 response during user lookup. Check your API key for profile '#{profile[:handle]}'"
    elsif response.code != "200"
      raise SlackMessage::ApiError, "Got an error back from the Slack API (HTTP #{response.code}):\n#{response.body}"
    elsif !(error.nil? || error == "")
      raise SlackMessage::ApiError, "Received error response from Slack during user lookup:\n#{response.body}"
    end
  end

  def self.blocks_message(blocks)
    if SlackMessage::Configuration.debugging?
      JSON.pretty_generate(blocks)
    else
      "[Enable debugging in configuration to view block data.]"
    end
  end
end
